Don't dispatch "pageclose" events if a "pageopen" wasn't dispatched for the page (PR 12747 follow-up)

Snuffleupagus opened the pull request:
Given that "pageopen" events are not guaranteed to occur, if the page becomes inactive *before* it finishes rendering, we should probably also avoid dispatching a "pageclose" event in that case to avoid confusing/inconsistent state in any event handlers.

Snuffleupagus committed:
Don't dispatch "pageclose" events if a "pageopen" wasn't dispatched for the page (PR 12747 follow-up)

Given that "pageopen" events are not guaranteed to occur, if the page becomes inactive *before* it finishes rendering, we should probably also avoid dispatching a "pageclose" event in that case to avoid confusing/inconsistent state in any event handlers.

Snuffleupagus commented:
/botio-linux preview

calixteman commented on web/base_viewer.js:
```diff
@@ -1517,6 +1517,9 @@ class BaseViewer {
     const { eventBus } = this;
 
     const dispatchPageClose = pageNumber => {
+      if (this._pageOpenPendingSet?.has(pageNumber)) {
```
The patch looks good.
An unrelated question: why do you lazily initialize `_pageOpenPendingSet` ?
Setting it in `_initializeScriptingEvent` will avoid few checks but maybe it's pretty rare to have a non null `_pageOpenPendingSet`.

Snuffleupagus replied:
> but maybe it's pretty rare to have a non null `_pageOpenPendingSet`.

That was my idea when writing the code, since that's what I observed in practice early during development.

However, to initialize things correctly in the general case, the final code also dispatches a "pageopen" event on the pre-existing "pagesinit" event and with those changes `_pageOpenPendingSet` is now always created. While the extra checks probably don't matter much, I'll fix this is a separate commit; thank you!

Snuffleupagus committed:
Unconditionally initialize the `this._pageOpenPendingSet` in `BaseViewer._initializeScriptingEvents` (PR 12747 follow-up)

With the code dispatching a "pageopen" event on the existing (general) `BaseViewer` event "pagesinit", in practice this means that the `Set` is always being created. Hence we can simplify the method overall, by always initializing the `this._pageOpenPendingSet` property.

timvandermeij commented:
/botio-linux preview

timvandermeij commented:
Thanks!
